shared/src/sources/balancer/swap/stable_math.rs

```diff
@@ -0,0 +1,485 @@
+use crate::sources::balancer::swap::fixed_point::Bfp;
+use crate::sources::balancer::swap::math::{rounded_div, BalU256};
+use ethcontract::U256;
+use lazy_static::lazy_static;
+
+lazy_static! {
+    static ref AMP_PRECISION: U256 = U256::from(1000);
+}
+
+fn calculate_invariant(
+    amplification_parameter: U256,
+    balances: Vec<Bfp>,
+    round_up: bool,
+) -> Result<U256, Error> {
+    let mut sum = U256::zero();
+    let num_tokens_usize = balances.len();
+    for balance_i in &balances {
+        sum = sum.badd(balance_i.as_uint256())?;
+    }
+    if sum.is_zero() {
+        return Ok(sum);
+    }
+
+    let mut invariant = sum;
+    let num_tokens = U256::from(num_tokens_usize);
+    let amp_times_total = amplification_parameter.bmul(num_tokens)?;
+    for _ in 0..255 {
+        let mut p_d = balances[0].as_uint256().bmul(num_tokens)?;
+        for balance in &balances[1..] {
+            // P_D = Math.div(Math.mul(Math.mul(P_D, balances[j]), numTokens), invariant, roundUp);
+            p_d = rounded_div(
+                p_d.bmul(balance.as_uint256())?.bmul(num_tokens)?,
+                invariant,
+                round_up,
+            )?
+        }
+        let prev_invariant = invariant;
+
+        invariant = rounded_div(
+            // invariant = Math.div(
+            //     Math.mul(Math.mul(numTokens, invariant), invariant).add(
+            //         Math.div(Math.mul(Math.mul(ampTimesTotal, sum), P_D), _AMP_PRECISION, roundUp)
+            //     ),
+            num_tokens
+                .bmul(invariant)?
+                .bmul(invariant)?
+                .badd(rounded_div(
+                    amp_times_total.bmul(sum)?.bmul(p_d)?,
+                    *AMP_PRECISION,
+                    round_up,
+                )?)?,
+            // Math.mul(numTokens + 1, invariant).add(
+            //     // No need to use checked arithmetic for the amp precision, the amp is guaranteed to be at least 1
+            //     Math.div(Math.mul(ampTimesTotal - _AMP_PRECISION, P_D), _AMP_PRECISION, !roundUp)
+            // ),
+            (num_tokens.badd(1.into())?)
+                .bmul(invariant)?
+                .badd(rounded_div(
+                    (amp_times_total.bsub(*AMP_PRECISION)?).bmul(p_d)?,
+                    *AMP_PRECISION,
+                    !round_up,
+                )?)?,
+            round_up,
+        )?;
+
+        match convergence_criteria(invariant, prev_invariant) {
+            None => continue,
+            Some(invariant) => return Ok(invariant),
+        }
+    }
+
+    Err(Error::StableInvariantDidntConverge)
+}
+
+fn calc_out_given_in(
+    amplification_parameter: U256,
+    mut balances: Vec<Bfp>,
+    token_index_in: usize,
+    token_index_out: usize,
+    token_amount_in: Bfp,
+) -> Result<Bfp, Error> {
+    let invariant = calculate_invariant(amplification_parameter, balances.clone(), true)?;
+    balances[token_index_in] = balances[token_index_in].add(token_amount_in)?;
```

Comment:
Whenever we index into balances like this (and in other functions to) that is a potential panic because the index can be out of bounds. Do we need to check this and return an error? I'm not sure.

Reply:
Yea, thats a pretty good point, I was literally mirroring the contract code, but we should use `array.get()?`

Reply:
It depends on whether it is possible to get the out of bounds error. Maybe that is supposed to already have been checked by a higher layer which calls these functions in which case panic would be fine.
If we do want to check it here yes you could use array.get .

Reply:
You can use a `array.get(index).ok_or(InvalidToken)`. They seem to have an error for this (although I'm not 100% sure its used for invalid indices in the pool tokens array:

https://github.com/balancer-labs/balancer-v2-monorepo/blob/a3875bdf1c5afb57092caf4d36062d906dcc2513/pkg/solidity-utils/contracts/helpers/BalancerErrors.sol#L126

Reply:
I'm finding it bit awkward to write the get_mut and update the value. Would it be appropriate to simply check first that `token_index_in < balances.len()` or is that not rusty enough?

Reply:
I have gone with this which I am not super happy with.

```rs
// Ensure no index error at token_index_in
let balance_token_in = balances.get(token_index_in).ok_or(Error::InvalidToken)?;
balances[token_index_in] = balance_token_in.add(token_amount_in)?;
```

Reply:
Checking `index < len` somewhere first is completely fine too. That  could be more ergonomic if you are querying the same index multiple times.

Reply:
Which of these functions would be public? If it isn't all then it would also be fine to only check bounds in those higher level functions while the others have the invariant that all indexes are valid.

Reply:
The only ones that will eventually be public are `calc_X_given_Y`.